packages/plugin-hardhat/src/prepare-upgrade.ts

```diff
@@ -1,18 +1,28 @@
 import { HardhatRuntimeEnvironment } from 'hardhat/types';
 import type { ContractFactory } from 'ethers';
 
-import { Options, ContractAddressOrInstance, deployImpl, getContractAddress } from './utils';
+import { Options, ContractAddressOrInstance, getContractAddress, deployProxyImpl, deployBeaconImpl } from './utils';
+import { getBeaconAddress, isBeaconProxy, isTransparentOrUUPSProxy } from '@openzeppelin/upgrades-core';
 
 export type PrepareUpgradeFunction = (
-  proxyAddress: ContractAddressOrInstance,
+  proxyOrBeaconAddress: ContractAddressOrInstance,
   ImplFactory: ContractFactory,
   opts?: Options,
 ) => Promise<string>;
 
 export function makePrepareUpgrade(hre: HardhatRuntimeEnvironment): PrepareUpgradeFunction {
-  return async function prepareUpgrade(proxy, ImplFactory, opts: Options = {}) {
-    const proxyAddress = getContractAddress(proxy);
-    const { impl } = await deployImpl(hre, ImplFactory, opts, proxyAddress);
-    return impl;
+  return async function prepareUpgrade(proxyOrBeacon, ImplFactory, opts: Options = {}) {
+    const proxyOrBeaconAddress = getContractAddress(proxyOrBeacon);
+    const { provider } = hre.network;
+    let deployedImpl;
+    if (await isTransparentOrUUPSProxy(provider, proxyOrBeaconAddress)) {
+      deployedImpl = await deployProxyImpl(hre, ImplFactory, opts, proxyOrBeaconAddress);
+    } else if (await isBeaconProxy(provider, proxyOrBeaconAddress)) {
+      const beaconAddress = await getBeaconAddress(provider, proxyOrBeaconAddress);
+      deployedImpl = await deployBeaconImpl(hre, ImplFactory, opts, beaconAddress);
+    } else {
```

Comment:
Added a check in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/456/commits/54297730341a409fd926abc8c13be51db9caa12f.

However, if the address has an `implementation()` function that returns something other than an address, the returned result may still look like an address according to `ethers.utils.isAddress()`.  In this case, the plugin would just return that the "address" is not found in the manifest.

This check is not necessary in `defender.proposeUpgrade` at this time, since https://github.com/ericglau/openzeppelin-upgrades/blob/beacon/packages/plugin-defender-hardhat/src/propose-upgrade.ts#L42 would give an error if the address is not a Transparent or UUPS proxy.  In order to add this check, we may need to do some refactoring similar to the suggestion in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/456#discussion_r768174558

Reply:
Refactored and added this check in `defender.proposeUpgrade()`